# Podstawowe typy obiektów dostępne w Pythonie

### Wypisywanie tekstu `print()`

W celu wypisania tekstu na ekranie należy skorzystać z komenty `print()`, gdzie w nawiasie wpisujemy tekst w podwójnym lub pojedynczym cudzysłowie `" "` lub `' '`.

In [1]:
print("Ala ma kota")

Ala ma kota


In [2]:
print('Ala ma kota')

Ala ma kota


Znaki specjalne `\n` znak nowej linii, `\t` znak tabulatury

In [3]:
print('Ala\nma\nkota')

Ala
ma
kota


In [4]:
print('Ala\tma\tkota')

Ala	ma	kota


In [5]:
print('Ala', 'ma', 'kota')

Ala ma kota


Argument separacji `sep` - dodatkowy argument w funkcji pozwalający na ustalenie jaki znak/znaki mają oddzielać kolejne elementy w funkcji; argument `end` - znak/znaki jakie mają wyświetlać się na końcu

In [6]:
print('Ala', 'ma', 'kota', sep='')

Alamakota


In [7]:
print('Ala', 'ma', 'kota', sep=' ')

Ala ma kota


In [8]:
print('Ala', 'ma', 'kota', sep='/')

Ala/ma/kota


In [9]:
print('Ala', 'ma', 'kota', sep='\n')

Ala
ma
kota


In [10]:
print('Ala', 'ma', 'kota', sep=' ', end='.')

Ala ma kota.

Uwaga! Podaczas pisania programów poza Jupyter Notebookiem oraz sesją interaktywną w celu wyświetlania wyniku działań, operacji itp. trzeba używać komendy `print()`. Np. `1+1` wpisane tutaj wyświetli `2`, ale w środowisku jak PyCharm nie wyświetli się nic i trzeba wpisać `print(1+1)`.

### Liczby

Liczyb całkowite `int`

In [11]:
1

1

In [12]:
type(1)

int

Liczby zmiennoprzecinkowe `float`

In [13]:
1.

1.0

In [14]:
type(1.)

float

Liczby zespolone `complex`

In [15]:
1+1j

(1+1j)

In [16]:
type(1+1j)

complex

### Operacje na liczbach

Dodawanie

In [17]:
2.43+4.54

6.970000000000001

Odejmowanie

In [18]:
2.43-5.45

-3.02

Mnożenie

In [19]:
2.43*4.54

11.032200000000001

Dzielenie liczb (różnice między zmiennoprzecinkowymi a całkowitymi)

In [20]:
4/3

1.3333333333333333

In [21]:
4//3

1

Dzielenie za pomocą `/` zawsze zwraca w wyniku liczbę zmiennoprzecinkową `float`.

In [22]:
type(4/3)

float

In [23]:
type(3/3)

float

Dzielenie za pomocą `//` zawsze zwraca w wyniku liczbę całkowitą `int`.

In [24]:
type(4//3)

int

Potęgowanie

In [25]:
2**3

8

Pierwiastkowanie

In [26]:
sqrt(4)

NameError: name 'sqrt' is not defined

Pierwiastokowanie nie jest zaimplementowane bezpośrednio w Pythonie, można jednak wykorzystać gotową **blibliotekę** zawierającą taką operację matematyczną 

In [28]:
from numpy import sqrt 

In [29]:
sqrt(4)

2.0

Uważniejsi mogli zauważyć, że wyniki dodawania i mnożenia niekoniecznie zgadzają się z rzeczywistością, sprawdźmy jeszcze raz

In [30]:
4.1*5.1

20.909999999999997

a powinno być `20.91`. Problem wynika z tego, że maszyna operuje na bitach, więc liczby są przekształcane na system binarny, a stąd "okrągłe" dla nas liczby w systemie dziesiętnym niekoniecznie takie są w ograniczonej pamięci komputera.

### Łańcuchy znaków

Łańcuchy znaków mają typ `str` od `string` i są zapisywane w cudzysłowie lub podwójnym cudzysłowie `' '`, `" "`

In [31]:
type('Laboratoria')

str

In [32]:
type("Laboratoria")

str

### Operacje na łańcuchach znaków

In [33]:
S='Laboratoria'

Długość możemy poznać za pomocą funkcji `len()`

In [34]:
len(S)

11

Wybieranie elementów z łańcucha odbywa się za pomocą *indeksów* nnumerujących każdy znak zaczynając **od zera**. Indeks podajemy w nawiasach kwadratowych `[]` 

In [35]:
S[0]

'L'

In [36]:
S[4]

'r'

Błąd przy wyjściu poza łańcuch

In [37]:
S[len(S)]

IndexError: string index out of range

In [38]:
S[len(S)-1]

'a'

Przedział znaków z określonym krokiem

In [39]:
S[1:10] #krok 1

'aboratori'

In [40]:
S[1:8:2] #krok 2

'aoao'

In [42]:
S[:5] # od początku do 4 indexu (bez 5)

'Labor'

In [41]:
S[5:] # od 5 indexu włącznie do końca

'atoria'

Indeksowanie od końca

In [43]:
S[-1]

'a'

In [44]:
S[-2]

'i'

In [45]:
S[:-2]

'Laborator'

Znajdowanie indeksu znaku za pomocą metody `find()`

In [46]:
S.find('a')

1

Kiedy znak nie zostanie znaleziony, zwrócona zostanie wartość `-1`

In [47]:
S.find('x')

-1

Konkatenacja, czyli składanie łańcuchów znaków

In [48]:
S1='Laboratoria'
S2='Python'

In [49]:
S1+S2

'LaboratoriaPython'

In [50]:
2*S1+S2

'LaboratoriaLaboratoriaPython'

Konwersja na wielkie i małe litery za pomocą metod `upper()` oraz `lower()`

In [51]:
S

'Laboratoria'

In [52]:
S.upper()

'LABORATORIA'

In [53]:
S.lower()

'laboratoria'

Podzielenie na ograniczniku na **listę** podłańcuchów za pomocą metody `split()`

In [54]:
S.split('o')

['Lab', 'rat', 'ria']

Łańcuchy znaków są **niezmienne**, co oznacza, że jeśli chcemy zmienić jedną literę na inną musimy stworzyć nowy łańcuch.

In [55]:
S[0]='B'

TypeError: 'str' object does not support item assignment

In [56]:
NewS='B'+S[1:]
NewS

'Baboratoria'